cpp/src/fil/cudf_to_matrix.cu

```diff
@@ -0,0 +1,151 @@
+#include "internal.cuh"
+#include <cstdint>
+#include <cudf/column/column_view.hpp>
+#include <cudf/types.hpp>
+#include <raft/handle.hpp>
+#include <thrust/host_vector.h>
+
+template <typename SrcT, typename ScaleT>
+__device__ float convert_maybe_check(const void* in,
+                                     std::size_t idx,
+                                     ScaleT scale,
+                                     int* lossy = nullptr)
+{
+  SrcT orig = reinterpret_cast<const SrcT*>(in)[idx];
+  float res = static_cast<double>(orig) * scale;
+  if (lossy != nullptr) {
+    SrcT back_cast = static_cast<double>(res) / scale;
+    if (back_cast != orig) atomicExch(lossy, 1);
+  }
+  return res;
+}
+
+__device__ float cast_scalar(const void* in, std::size_t idx, cudf::data_type from, int* lossy)
+{
+  switch (from.id()) {
+    // below, we're checking for rounding. _typecast_will_lose_information() only checks for
+    // out-of-range, which could only happen in FLOAT64 here.
```

Comment:
What about `[u]int{32,64}`, for values that aren't exactly representable as `float32`?